README
=========   ---------------------------
 CL-WIKI    Wiki Engine for Common Lisp
=========   ---------------------------

This is version 0.4.x alpha of cl-wiki. It is incompatible with previous 
0.3.x versions of cl wiki. Markup of pages was totally changed.

Dependencies
============

It uses Hunchentoot (1.2.x), CL-PPCRE, CL-WHO (min. 0.6.0), 
and CL-EMB (min. 0.4.2), colorize – all of them available in quicklisp.

Quick start
===========

        (asdf:operate 'asdf:load-op :cl-wiki)
        ;; Write config file "wiki.conf" in current directory with at least
        ;; (base (directory "/path/to/wiki-data/"))
        ;; ...
        (wiki:start)

or using quicklisp:
   
	(push #p"your_path_to_clwiki_code/cl-wiki/" asdf:*central-registry*)
	(ql:quickload :cl-wiki)
        (wiki:start #p"your_path_to_clwiki_config/wiki.conf")


Then direct your browser to http://localhost:5757/Somepage


CL-WIKI uses Hunchentoot <http://weitz.de/hunchentoot/>. The example above
runs it without any front-end. But you can install it for use with Apache and
mod_lisp, too. See Hunchentoot's documentation.

Syntax highlighting via https://github.com/redline6561/colorize loaded via quicklisp.org

Config
======

File "wiki.conf" in current directory or in path given as parameter to `wiki:start`.
Example:

  ;; Config file for CL-WIKI -- Wiki engine for Common Lisp
  (base (port 5757
         directory "/home/stesch/wiki-data"
         url "/cl-wiki/"
         page-list ((title "Home"    link "Home")
                    (title "About"   link "About")
                    (title "Contact" link "Contact"))))


The following parameters can be set with the config file:

Section: base
-------------
directory       Base directory for the wiki data, like pages, etc.

port            TCP/IP port CL-WIKI will listen to. Attention: listens
                on every interface.

template-directory      Base directory for templates.

page-list       List of title-link pairs with links that get displayed on
                the top and bottom of every page. See example above how to
                set this.

reverse-proxy   Indicates if CL-WIKI runs behind a reverse proxy.
                Default is NIL. Set it to t when behind a proxy.

home-page       Default page when root directory of the site is accessed.

edit-textarea-size      Plist of size of edit textarea.
                        Default is (:rows 25 :cols 80).


Textformatting
==============

Content of page is written as s-expressions. Symbols can be written with 
or without colon. They are interned in „keywords” package. In all places 
where is "text" there can be more of them and they can be mixed with markup, 
they are all concatenated.

(:wiki "name")                  ==> Generates a link to named page.
(:wiki "name" "text")           ==> Generates a link to named page
                                    and links Text.
(:link "url" "text")            ==> Inserts external link
(:em "text")                    ==> emphasize text (<em>text</em>) 
                                    -- on most browsers italic
(:strong "text")                ==> stronger emphasize text
                                    (<strong>text</strong>)
                                    -- on most browsers bold
(:sub "text")                   ==> subscripting x
(:sup "text")                   ==> superscripting x
(:s "text")                     ==> strike out text
(:u "text")                     ==> underline text
(:del "text")                   ==> Mark text as deleted
(:ins "text")                   ==> Mark text as inserted
(:br)                           ==> Start a new line
(:hr)                           ==> Horizontal line

(ul ...), (ol ...), (li ...)    ==> Unordered and ordered list, list item

(section ...)                   ==> No visual rendering, increments header level
(title "text")                  ==> Header, level dependent of sections nesting level

(source "lang" "source")        ==> Sourcecode snippet, colorized for the
                                    named language LANG. For now its necessary to manually 
               			    escape " inside. Snippet is surrounded by
                                    <div class="code"> ... </div>
                                    LANG must be one of (taken from colorize README):
				        Common Lisp,
					Emacs Lisp,
    					Scheme,
					Clojure, [alpha]
    					C,
    					C++,
    					Java,
    					Python,
    					Erlang,
    					Haskell,
    					Objective-C,
    					Diff,
    					Webkit.
				    If lang code is unrecognized, then no coloring is used.

Unrecognized tags are rendered as text. They must be well formed forms.

Philosophy
==========

The term "wiki wiki" and its meaning are well known. Quick. Quick to use
and quick to program. The first wiki was written in 30 lines of Perl
code.


Thanks
======
Kevin Griffin for tests with different CL implementations and
beeing an early adopter.
Ian Clelland for ideas, patches, etc.
Vehbi Sinan Tunalioglu for ideas, patches, etc.
Jens Teich reviving development, ideas, patches, etc.
Ralf Stoye for a patch to speedup regexps, which are heavily used in CL-WIKI.


CHANGELOG

0.4.0 alpha
	2014-05-16
	New markup based on s-expressions. Previous markup disabled.
	Whole system incompatible with previous wersions.

	2014-05-14
	Fix: uncoment temporarily commented and forgotten
	(setf emb:*locking-function* 'emb-lock-function)

0.3.4
	2014-05-13 (Piotr Chamera)
	`cl-wiki:start` now takes optional parameter – path to wiki 
	configuration file. Default is „wiki.conf” in current directory 
	(as before).

	Removed local copy of colorize code. Now depends on colorize
	library available from quicklisp.org.

	Removed aif macro (and all uses of it from code).

0.3.3
	2014-05-13 (Piotr Chamera)
	Modified to work with current quicklisp release of libraries.
	Added sample „wiki.conf” file.

0.3.2
	2008-11-19
	Regexp/rendering speedup. Thanks to Ralf Stoye.

0.3.1
	2008-07-07
	New codes <small>, <big>, <s>, <u>, <del>, <ins>, <br>, <sub>, <sup>, 
	comments, droping dependency on kmrcl

0.3.0
	2008-06-20
	New code <nowiki>, preview of edited page

0.2.4
	2008-06-16
	Small fix to please SBCL, start script for SBCL

0.2.3
	2008-06-15
	Source snippets colorized with colorize code from lisppaste


util.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-

;;; This software is Copyright (c) Stefan Scholl, 2005-2008.
;;; Modified by Piotr Chamera, 2014.
;;; Stefan Scholl grants you the rights to distribute
;;; and use this software as governed by the terms
;;; of the Lisp Lesser GNU Public License
;;; (http://opensource.franz.com/preamble.html),
;;; known as the LLGPL.

(in-package :cl-wiki)

;; -----------------------------------------------------------------------------
;; Macros
;; -----------------------------------------------------------------------------

;; (defmacro aif (test then &optional else)
;;   "Like IF. But the value of the test form gets bound to IT."
;;   `(let ((it ,test))
;;     (if it ,then ,else)))


;; -----------------------------------------------------------------------------
;; Time functions
;; -----------------------------------------------------------------------------

(defun iso-date-time (&optional (time (get-universal-time)))
  "Returns string with date + time according to ISO 8601."
  (multiple-value-bind
        (second minute hour date month year)
      (decode-universal-time time 0)
    (format nil "~4,'0D-~2,'0D-~2,'0D ~2,'0D:~2,'0D:~2,'0D"
            year month date hour minute second)))


codes.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-

;;; This software is Copyright (c) Stefan Scholl, 2005-2008.
;;; Modified by Piotr Chamera, 2014.
;;; Stefan Scholl grants you the rights to distribute
;;; and use this software as governed by the terms
;;; of the Lisp Lesser GNU Public License
;;; (http://opensource.franz.com/preamble.html),
;;; known as the LLGPL.

(in-package :cl-wiki)

;; Handling wiki codes


(defparameter *marker* (map 'string #'code-char #(#x1e #xff #xfe #x1e))
  "To mark the place in the string where preserved HTML should be restored.")

(defun rep-url-encoder-1 (target-string start end match-start match-end
                          reg-starts reg-ends)
  "Replacement function for CL-PPCRE:REGEX-REPLACE-ALL. Match in register 1 (0 in array)
gets url encoded."
  (declare (ignore start end match-start match-end))
  (tbnl:url-encode (subseq target-string (aref reg-starts 0) (aref reg-ends 0))))

(defun rep-nowiki (target-string start end match-start match-end
                   reg-starts reg-ends)
  "Replacement function for CL-PPCRE:REGEX-REPLACE-ALL. Match in register 1 (0 in array)
will be escaped and stored in *preserved-html*."
  (declare (special *preserved-html*) (ignore start end match-start match-end))
  (prog1
      (format nil "~A~D~A" *marker* (fill-pointer *preserved-html*) *marker*)
    (vector-push-extend
     (escape-string (subseq target-string (or (aref reg-starts 0) 0) (or (aref reg-ends 0) 0)))
     *preserved-html*)))

(defun rep-list (target-string start end match-start match-end reg-starts reg-ends)
  "Replacement function for CL-PPCRE:REGEX-REPLACE-ALL. Handles lists."
  (declare (ignore start end reg-starts reg-ends))
  (with-output-to-string (s)
    (write-string "<ul>" s)
    (cl-ppcre:do-matches-as-strings (line "(?ims)(?:<br>)?.*?(?=(?:<br>|\\z))"
                                          (subseq target-string match-start match-end)
                                          nil :sharedp t)
      (write-string (ppcre:regex-replace "(?:<br>|^)\\*\\s*(.*)" line "<li>\\1</li>") s))
    (write-string "</ul>" s)))

(defun make-source-snippet (lang string)
  "Make a source snippet (colorized or just plain PRE) out of string."
  (if lang
      (concatenate 'string "<div class=\"code\">" (colorize:html-colorization lang string 'colorize::encode-for-tt) "</div>")
      (concatenate 'string "<pre class=\"code\">" string "</pre>")))

(defun rep-source (target-string start end match-start match-end
                   reg-starts reg-ends)
  "Replacement function for CL-PPCRE:REGEX-REPLACE-ALL. Match in register 1 (0 in array)
is the language, match in register 2 becomes a source snippet. Stored in *preserved-html*."
  (declare (special *preserved-html*) (ignore start end match-start match-end))
  (prog1
      (format nil "~A~D~A" *marker* (fill-pointer *preserved-html*) *marker*)
    (vector-push-extend
     (let* ((lang-string (subseq target-string (or (aref reg-starts 0) 0) (or (aref reg-ends 0) 0)))
            (lang (first (rassoc lang-string (colorize:coloring-types) :test #'string-equal))))
       (make-source-snippet lang (subseq target-string (aref reg-starts 1) (aref reg-ends 1))))
     *preserved-html*)))

(defun rep-restore-preserved (target-string start end match-start match-end
                              reg-starts reg-ends)
  "Replacemen function for CL-PPCRE:REGEX-REPLACE-ALL. Match in register 1 (0 in array)
is an index for the array *preserved-html*. Restores preserved HTML."
  (declare (special *preserved-html*) (ignore start end match-start match-end))
  (aref *preserved-html* (parse-integer (subseq target-string (aref reg-starts 0) (aref reg-ends 0)))))
  
(defparameter *unescaped-replace-list* ; re-evaluate *unescaped-replace-list-compiled* if you change this
  `(
    ;; Remove the marker first, so there will be no trouble at the end.
    (,*marker* . "")

    ;; <nowiki> .. </nowiki> ==> Exclude portions of text from wiki converting
    ("(?ims)<nowiki>(.*?)</nowiki>" . ,'rep-nowiki)

    ;; <!-- comment --> ==> Removes everything between <!-- and -->. Allows commenting.
    ("(?ims)<!--(.*?)-->" . "")

    ;; <source> .. </source> ==> Sourcecode
    ("(?ims)<source(?:\\s*lang=\"(.*?)\")?>(?:\\x0d?\\x0a)*(.*?)</source>" . ,'rep-source))
  "List of conses. CAR is regexp and CDR replace-string for
CL-PPCRE:REGEX-REPLACE-ALL. This list is used before string gets escaped
for HTML.")

(defparameter *replace-list* ; re-evaluate *replace-list-compiled* if you change this
  `(
    ;; CR or LF ==> <br>
    ("\\x0d?\\x0a" . "<br>")

    ;; <br> or <br/> ==> Start a new line
    ("&lt;br\s*/?&gt;(<br>)?" . "<br/>")

    ;; '''text''' ==> stronger emphasize text (<strong>text</strong>),
    ;;                on most browsers bold.
    ("'''(.*?)'''" . "<strong>\\1</strong>")

    ;; ''text'' ==> emphasize text (<em>text</em>), on most browsers italic.
    ("''(.*?)''" . "<em>\\1</em>")

    ;; <small>text</small> ==> text in small font
    ("&lt;small&gt;(.*?)&lt;/small&gt;" . "<small>\\1</small>")

    ;; <big>text</big> ==> text in big font
    ("&lt;big&gt;(.*?)&lt;/big&gt;" . "<big>\\1</big>")

    ;; <sub>x</sub> ==> subscripting x
    ("&lt;sub&gt;(.*?)&lt;/sub&gt;" . "<sub>\\1</sub>")

    ;; <sup>x</sup> ==> superscripting x
    ("&lt;sup&gt;(.*?)&lt;/sup&gt;" . "<sup>\\1</sup>")

    ;; <s>text</s> ==> strike out text
    ("&lt;s&gt;(.*?)&lt;/s&gt;" . "<s>\\1</s>")

    ;; <u>text</u> ==> underline text
    ("&lt;u&gt;(.*?)&lt;/u&gt;" . "<u>\\1</u>")

    ;; <del>text</del> ==> Mark text as deleted
    ("&lt;del&gt;(.*?)&lt;/del&gt;" . "<del>\\1</del>")

    ;; <ins>text</ins> ==> Mark text as inserted
    ("&lt;ins&gt;(.*?)&lt;/ins&gt;" . "<ins>\\1</ins>")

    ;; [[Page|Text]] ==> Generates a link to named page and links Text.
    ("\\[\\[([^]\">|]*?)\\|([^]\">]*?)\\]\\]"
     . ,(list "<a href=\"" 'rep-url-encoder-1 "\" title=\"" 0 "\">" 1 "</a>"))

    ;; [[Page]] ==> Generates a link to named page.
    ("\\[\\[(.*?)\\]\\]"
     . ,(list "<a href=\"" 'rep-url-encoder-1 "\" title=\"" 0 "\">" 0 "</a>"))

    ;; [http://www.somepage.example/] ==> Inserts external link.
    ("\\[(http|https|ftp|mailto|gopher):([^\">].*?)\\]"
     . "<a rel=\"nofollow\" class=\"external \\1\" href=\"\\1:\\2\" title=\"\\1:\\2\">\\1:\\2</a>")

    ;; ====== Foo ======  ==> Level 6 header (etc.)
    ("(?ims)(<br>|^)======\\s*(.*?)\\s*=*(?=(?:<br>|\\z))" . "\\1<h6>\\2</h6>")
    ("(?ims)(<br>|^)=====\\s*(.*?)\\s*=*(?=(?:<br>|\\z))" . "\\1<h5>\\2</h5>")
    ("(?ims)(<br>|^)====\\s*(.*?)\\s*=*(?=(?:<br>|\\z))" . "\\1<h4>\\2</h4>")
    ("(?ims)(<br>|^)===\\s*(.*?)\\s*=*(?=(?:<br>|\\z))" . "\\1<h3>\\2</h3>")
    ("(?ims)(<br>|^)==\\s*(.*?)\\s*=*(?=(?:<br>|\\z))" . "\\1<h2>\\2</h2>")
    ("(?ims)(<br>|^)=\\s*(.*?)\\s*=*(?=(?:<br>|\\z))" . "\\1<h1>\\2</h1>")

    ;; ---- ==> Horizontal divider
    ("(?ims)(<br>|^)----.*?(?=(?:<br>|\\z))" . "\\1<hr>")

    ;; * Item ==> List item
    ("(?ims)(?:<br>|^)\\*.*?(?=(?:<br>|^)[^\\*]|\\z)" . ,'rep-list)

    ;; <br><br><br>* ==> <p>
    ("(?ims)(<br>){2,}" . "<p>")

    ;; <br> => " "
    ("(?ims)(<br>)" . " ")

    ;; <br/> => "<br>" - Cleanup for <br> wiki tag
    ("(?ims)<br/>" . "<br>")


    ;; Restore preserved HTML
    (,(concatenate 'string *marker* "(\\d+)" *marker*) . ,'rep-restore-preserved)
    )
  ;;
  "List of conses. CAR is regexp and CDR replace-string for
CL-PPCRE:REGEX-REPLACE-ALL.")

(defparameter *unescaped-replace-list-compiled*
  (mapcar #'(lambda (reg-rep)
              (destructuring-bind (regexp . replace) reg-rep
                (cons (cl-ppcre:create-scanner regexp) replace)))
          *unescaped-replace-list*))

(defparameter *replace-list-compiled*
  (mapcar #'(lambda (reg-rep)
              (destructuring-bind (regexp . replace) reg-rep
                (cons (cl-ppcre:create-scanner regexp) replace)))
          *replace-list*))

(defun translate-wiki-codes (string)
   "Translate wiki codes."
   (let ((*preserved-html* (make-array 3 :element-type 'string
                                       :adjustable t :fill-pointer 0)))
     (declare (special *preserved-html*)) ; To temporary store HTML code ;
    ;; First with the unescaped string  ;
     (loop for (regexp . replace) in *unescaped-replace-list-compiled*
        do (setf string (ppcre:regex-replace-all regexp string replace)))
    ;; Replacement in the escaped string ;
     (setf string (escape-string string))
     (loop for (regexp . replace) in *replace-list-compiled*
        do (setf string (ppcre:regex-replace-all regexp string replace)))
     string))


(defun translate (string)
  "Translates the wiki codes inside STRING into HTML."
  (translate-wiki-codes string))


codes1.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-

;;; This software is Copyright (c) Piotr Chamera, 2014.
;;; 
;;; Lisp Lesser GNU Public License
;;; (http://opensource.franz.com/preamble.html),
;;; known as the LLGPL.

;; Handling wiki text

(defpackage #:cl-wiki.sxml-text
  (:nicknames #:st)
  (:use #:cl #:colorize)
  (:export #:translate-wiki-codes))

(in-package :cl-wiki.sxml-text)

(defun con (&rest args)
  "Concatenate args into string."
  (apply #'concatenate 'string args))

(defun wrap-into-tag (tag in)
  "Wraps processed children into html tag."
  (con "<" tag ">" (process-list in) "</" tag ">"))

(defun empty-tag (tag)
  "Generate empty html tag."
  (con "<" tag "/>"))


(defvar *test-structure* nil
  "List structure for testing.")
(setf *test-structure*
      '(:section 
	(:title "test tytułu" " ABC.") 
	(:p "To jest " (:em "akapit") " z " (:link "linkiem") ".")
	(:p "Drugi akapit e = m c" (:^ "2") ".")))

(defvar *section-level* 1
  "Level of section nesting. Top level is 1, and up for next levels.")

(defvar *page-source* nil
  "Source tree of whole page.")


(defun process-list (in)
  "Process list of nodes."
  (apply #'concatenate 'string (mapcar #'process-child in)))


(defun process-child (in)
  "Process single node."
  (cond ((stringp in)
	 in)
	((listp in)
	 (let ((name (first in))
	       (children (rest in)))
	   (case name
	     (:section (section children))
	     (:title   (title children))
	     (:source  (source children))
	     (:p       (wrap-into-tag "p" children))
	     (:ul      (wrap-into-tag "ul" children))
	     (:ol      (wrap-into-tag "ol" children))
	     (:li      (wrap-into-tag "li" children))
	     (:em      (wrap-into-tag "em" children))
	     (:strong  (wrap-into-tag "strong" children))
	     (:sup     (wrap-into-tag "sup" children))  ;; superscript
	     (:sub     (wrap-into-tag "sub" children))  ;; subscript
	     (:s       (wrap-into-tag "s" children))    ;; strike out text
	     (:u       (wrap-into-tag "u" children))    ;; underline text
	     (:del     (wrap-into-tag "del" children))  ;; mark text as deleted
	     (:ins     (wrap-into-tag "ins" children))  ;; mark text as inserted
	     (:hr     (empty-tag "hr"))
	     (:br     (empty-tag "br"))
	     (:link   (link-external children))
	     (:wiki   (link-wiki children))
	     (:page-source (page-source))
	     (otherwise (format nil "~s" in)))))))
	 

(defun section (in)
  (let ((*section-level* (1+ *section-level*)))
    (process-list in)))
  
(defun title (in)
  (let ((tag (case *section-level*
	       (1 "h1")
	       (2 "h2")
	       (3 "h3")
	       (4 "h4")
	       (5 "h5")
	       (6 "h6")
	       (T "h6"))))
    (con "<" tag ">" (process-list in) "</" tag ">")))
 
(defun source (in)
  (let* ((lang-name (first in))
	 (lang (first (rassoc lang-name (colorize:coloring-types) :test #'string-equal)))
	 (string (apply #'concatenate 'string (rest in))))
    (if lang
	(concatenate 'string "<div class=\"code\">" (colorize:html-colorization lang string 'colorize::encode-for-tt) "</div>")
	(concatenate 'string "<pre class=\"code\">" string "</pre>"))))

(defun page-source ()
  (let* ((lang (first (rassoc "Common Lisp" (colorize:coloring-types) :test #'string-equal)))
	 (string (format nil "~a" *page-source*)))
    (concatenate 'string "<div class=\"code\">" 
		 (colorize:html-colorization lang string 'colorize::encode-for-tt) 
		 "</div>")))

(defun link-external (in)
  (let* ((url (first in))
	 (text (if (rest in)
		   (process-list (rest in))
		   (first in))))
    (con "<a href=\"" url "\">" text "</a>")))

(defun link-wiki (in)
  (let* ((name (first in))
	 (text (if (rest in)
		   (process-list (rest in))
		   (first in))))
    (con "<a href=\"" (tbnl:url-encode name) "\">" text "</a>")))


(defun translate-wiki-codes (string)
   "Translate wiki codes."
   (let ((*read-eval* nil)
	 (*package* (find-package :keyword))
	 (*page-source* string))
     (labels ((read-objects (string &optional (pos 0) (acc nil))
		"Read sequence of objects from string."
		(multiple-value-bind (object last-position)
		    (read-from-string string nil nil :start pos)
		  (if object
		      (progn 
			(push object acc)
			(read-objects string last-position acc))
		      (progn
			(push (subseq string last-position (length string)) acc)
			(reverse acc))))))
       (process-list (read-objects string)))))

;(in-package :cl-wiki)

(defun translate (string)
  "Translates the wiki codes inside STRING into HTML."
  (cl-wiki.sxml-text:translate-wiki-codes string))


wiki.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-

;;; This software is Copyright (c) Stefan Scholl, 2005-2008.
;;; Modified by Piotr Chamera, 2014.
;;; Stefan Scholl grants you the rights to distribute
;;; and use this software as governed by the terms
;;; of the Lisp Lesser GNU Public License
;;; (http://opensource.franz.com/preamble.html),
;;; known as the LLGPL.

(in-package :cl-wiki)

(defvar *wiki-port* 5757
  "Port CL-WIKI listens for http connections.")

(defvar *wiki-directory* nil
  "Base directory for data.")

(defvar *wiki-template-directory*
  (merge-pathnames "templates/"
                   (asdf:component-pathname (asdf:find-system :cl-wiki)))
  "Base directory for templates.")

(defvar *wiki-page-list* '((:title "Home" :link "Home")
                           (:title "About" :link "About")
                           (:title "Contact" :link "Contact"))
  "Plist of pages displayed on every page. :LINK can have parameters.")

(defvar *wiki-home-page* "Home"
  "Default page when root directory of the site is accessed.")

(defvar *behind-proxy-p* nil
  "Indicates if CL-WIKI runs behind a reverse proxy.")

(defvar *page-index* (make-hash-table :test #'equal)
  "Global index over all wiki pages.")

(defvar *edit-textarea-size* '(:rows 25 :cols 80)
  "List of two integer values containing number of rows and columns of textarea
of edit form.")

;; -------------------------------------------------------------------

(defvar *wiki-server* nil
  "Hunchentoot server instance")

(defvar *emb-lock* (bordeaux-threads:make-lock "emb-lock")
  "Lock for CL-EMB.")

(defvar *page-index-lock* (bordeaux-threads:make-lock "page-index-lock")
  "Lock for *PAGE-INDEX*")

;; -------------------------------------------------------------------

(defun current-page ()
  "Extracts the current page from the URL. 
Returns *WIKI-HOME-PAGE* when / is accessed."
  (let ((url (hunchentoot:script-name*)))
    (hunchentoot:url-decode (if (string= "/" url)
                          *wiki-home-page*
                          (subseq url (+ 1 (position #\/ url :from-end t)))))))

(defun page-version (page)
  "Current version of PAGE in *PAGE-INDEX*"
  (bordeaux-threads:with-lock-held (*page-index-lock*)
    (gethash page *page-index*)))


(defun contents-of-file (pathname)
  "Returns a string with the entire contents of the specified file."
  (with-open-file (in pathname :direction :input)
    (contents-of-stream in)))

;; From lemonodor's LSP
(defun contents-of-stream (stream)
  "Returns a string with the entire contents of the specified stream."
  (with-output-to-string (contents)
    (let* ((buffer-size 4096)
           (buffer (make-string buffer-size)))
      (labels ((read-chunks ()
                 (let ((size (read-sequence buffer stream)))
                   (if (< size buffer-size)
                       (princ (subseq buffer 0 size) contents)
                       (progn
                         (princ buffer contents)
                         (read-chunks))))))
        (read-chunks)))))

(defun meta (pathname)
  "Reads and returns meta data of the page with given PATHNAME."
  (with-open-file (in pathname :direction :input)
    (read in)))

(defun meta-and-content (pathname)
  "Reads and returns meta data and content of the page with given PATHNAME."
  (with-open-file (in pathname :direction :input)
    (values (read in)
            (contents-of-stream in))))

(defun ncomplete-meta (meta)
  "Completes the META data with a few additional values such as the string
at :LAST-MODIFIED, which is the ISO 8601 representation of the universaltime at
:TIME."
  (nconc meta (list :last-modified (iso-date-time (getf meta :time)))))

(defun index-path ()
  "Filesystem path to page index file."
  (merge-pathnames (make-pathname :name "all" :type "index") *wiki-directory*))

(defun set-page-version (page version)
  "Sets the VERSION of given PAGE in the *PAGE-INDEX*. Don't forget to lock!"
  (setf (gethash page *page-index*) version)
  (with-open-file (out (index-path) :direction :output :if-exists :supersede)
    (loop for key being the hash-keys in *page-index* using (hash-value value)
          do (format out "~A~%~A~%" key value))))

(defun save (page)
  "Save the submitted PAGE and display it."
  (bordeaux-threads:with-lock-held (*page-index-lock*)
    (let* ((version (+ 1 (gethash page *page-index* 0))) ; if page not in index then version = 1, 
					                 ; else version = current version + 1
           (path (page-version-path page version)))
      (ensure-directories-exist path)
      (with-open-file (out path :direction :output :if-exists :supersede)
        (prin1 (list :address (if *behind-proxy-p*
                                  (hunchentoot:header-in* "X-Forwarded-For")
                                  (hunchentoot:remote-addr*))
                     :user nil      ; XXX just a test value
                     :time (get-universal-time))
               out)
        (write-string (hunchentoot:parameter "content") out))
      (set-page-version page version)))
  (display page (page-version page)))

(defun page-version-path (page version)
  "Filesystem path for given PAGE and VERSION number."
  (merge-pathnames (make-pathname :directory (list :relative page)
                                  :name (princ-to-string version)
                                  :type "page")
                   *wiki-directory*))

(defun page-path (page version)
  "Filesystem path for given PAGE. NIL if non existant."
  (let ((path (page-version-path page version)))
    (when (probe-file path)
      path)))
  
(defun execute-main-template (page version body &key edit meta)
  "Execute the main template with all data needed. EDIT is t if you display the
edit form. The plist META contains the meta data like :TIME for last modified time."
  (emb:execute-emb (merge-pathnames "main.template" *wiki-template-directory*)
                   :env `(:edit ,edit
                          :page-list ,*wiki-page-list*
                          :edit-url ,(format nil "~A?version=~D&action=edit"
                                             page version)
                          :prev-link ,(when (and version (> version 1))
                                            (format nil "~A?version=~D"
                                                    page (1- version)))
                          :version ,(princ-to-string version)
                          :title ,page
                          :page ,page
                          :meta ,meta
                          :body ,body)))

(defun edit (page version &key preview)
  "Edit the PAGE."
  (multiple-value-bind (meta content)
      (let ((path (page-path page version)))
	(if path
           (if preview
               (values (meta path) (hunchentoot:post-parameter "content"))
               (meta-and-content path))
           (values nil 
                   (if preview
                       (hunchentoot:post-parameter "content")
                       (format nil "Describe ~A here." page))))); XXX All texts must be configurable. L10N.
    (let ((body (with-html-output-to-string (s)
                  (when preview
                    (htm 
                     (:h2 "Preview")
                     (:hr)
                     (:strong "This is only a preview! Changes aren't saved!")
                     (:hr)
                     (str (cl-wiki.sxml-text::translate content))
                     ;(str (translate content))
                     (:hr)))
                  (:form :action (format nil "~A" page) :method "POST"
                         (:fieldset
                          (:legend "Edit page")
                          (:p
                           (:textarea :name "content"
                                      :rows (getf *edit-textarea-size* :rows)
                                      :cols (getf *edit-textarea-size* :cols)
                                      (esc content)))
                          (:input :type "submit" :name "action-preview" :value "Preview")
                          (str " ")
                          (:input :type "submit" :name "action-save" :value "Save"))))))
      (execute-main-template page version body :edit t :meta meta))))

(defun display (page version)
  "Display PAGE."
  (let ((path (page-path page version)))
    (if path
        (multiple-value-bind (meta content)
            (meta-and-content path)
          (execute-main-template page version (cl-wiki.sxml-text::translate content) :meta (ncomplete-meta meta)))
          ;(execute-main-template page version (translate content) :meta (ncomplete-meta meta)))
        (edit page version))))

(defun debug-header ()
  "Shows the request header for debugging purpose."
  (with-html-output-to-string (s)
    (:html (:head (:title "Request header - DEBUG - CL-WIKI"))
           (:body (:h1 "Request header")
                  (:table :border 1
                   (:tr
                    (:th "Name") (:th "Value"))
                   (loop for (name . value) in (hunchentoot:headers-in*)
                         do (htm (:tr (:td (str name)) (:td (str value))))))))))

(defun wiki ()
  "Main command dispatching function for CL-WIKI."
  (let* ((action (hunchentoot:parameter "action"))
         (action-save (hunchentoot:parameter "action-save"))
         (action-preview (hunchentoot:parameter "action-preview"))
         (page (current-page))
         (max-version (page-version page))
         (param-version (let ((it (hunchentoot:parameter "version")))
			  (when it (parse-integer it :junk-allowed t))))
         (version (if (and param-version (< 0 param-version max-version))
                      param-version
                      max-version)))
    (cond
      ((string= "debug-header" action) (debug-header))
      ((or (string= "edit" action) action-preview) (edit page version :preview action-preview))
      (action-save (save page))
      (t (display page version)))))

(defun emb-lock-function (func)
  "Lock function for CL-EMB."
  (bordeaux-threads:with-lock-held (*emb-lock*)
    (funcall func)))

;; -------------------------------------------------------------------

(defun read-config (path)
  (with-open-file (plist-stream path)
    (let ((*read-eval* nil)
          (*package* (find-package :keyword)))
      (read plist-stream))))

(defun get-config-value (config section key &optional default)
  (getf (getf config section) key default))

;; Partially taken from PCL 
(defun string-to-directory-pathname (string)
  (when string
    (let ((pathname (pathname string)))
      (make-pathname 
       :directory (append (or (pathname-directory pathname) (list :relative))
                          (list (file-namestring pathname)))
       :name      nil
       :type      nil
       :defaults pathname))))

(defun read-page-index ()
  "Filling *PAGE-INDEX*."
  (bordeaux-threads:with-lock-held (*page-index-lock*)
    (clrhash *page-index*)
    (with-open-file (in (index-path) :direction :input :if-does-not-exist nil)
      (when in
        (loop for key = (read-line in nil nil)
              while key
              for value = (parse-integer (read-line in t))
              do (setf (gethash key *page-index*) value))))))

(defun init (config-path)
  (let ((conf (read-config config-path)))
    (setf *wiki-directory* (string-to-directory-pathname (get-config-value conf :base :directory))
          *wiki-port* (get-config-value conf :base :port *wiki-port*)
          *wiki-template-directory* (get-config-value conf :base :template-directory *wiki-template-directory*)
          *wiki-page-list* (get-config-value conf :base :page-list *wiki-page-list*)
          *wiki-home-page* (get-config-value conf :base :home-page *wiki-home-page*)
          *behind-proxy-p* (get-config-value conf :base :reverse-proxy *behind-proxy-p*)
          *edit-textarea-size* (get-config-value conf :base :edit-textarea-size *edit-textarea-size*)))

  (read-page-index)

  (setf *attribute-quote-char* #\"
        (html-mode) :sgml
        *escape-char-p* #'(lambda (c) (find c "<>&"))
        emb:*locking-function* 'emb-lock-function
        hunchentoot:*default-content-type* "text/html; charset=utf-8")

  (push (hunchentoot:create-prefix-dispatcher "/" 'wiki)
	hunchentoot:*dispatch-table*)

  (push (hunchentoot:create-folder-dispatcher-and-handler
	 "/_static/" (merge-pathnames "_static/"
				      wiki::*wiki-directory*))
	hunchentoot:*dispatch-table*)
  )

;; -------------------------------------------------------------------

(defun start (&optional config-path)
  (init (or config-path
	    (make-pathname :name "wiki" :type "conf")))
  (unless *wiki-directory*
    (error "A directory must be specified. Update your config, please."))
  (setf *wiki-server* (hunchentoot:start (make-instance 'hunchentoot:easy-acceptor 
							:port *wiki-port*))))

(defun stop ()
  (hunchentoot:stop *wiki-server*))
